src/lib.rs

```diff
@@ -360,6 +463,11 @@ impl<'m> Migrations<'m> {
             let m = &self.ms[v];
             if let Some(ref down) = m.down {
                 debug!("Running: {}", down);
+
+                if let Some(hook) = &m.down_hook {
+                    hook(&tx);
```

Comment:
Regarding the type passed to the hook here, I’ve fiddled with this more and `&Transaction` seems the viable option.
We can’t pass:
* `&Connection` or `&mut Connection` because the connection is already mut borrowed to create the transaction
* `&mut Transaction` because of https://github.com/rust-lang/rust/issues/57349 (this occurs in the `Migration` struct, on the hook fields with an `Option` type).

That’s probably fine, if someone needs multiple transactions for a hook, they can always create multiple migrations with hooks and each will run in their own transaction.

Also, the hook could return a `Result`, and if it’s `Result::Err`, the transaction would be aborted (this would be used with hooks performing consistency checks before running a downward migration for instance.). It’s not necessary to implement this in this PR, but in general, what do you think @matze ?

Reply:
I like that idea and added a rough draft but I am not convinced it's the nicest solution with two-level Results to be honest. But I don't see how else to do it nicely.

Sorry about the delay, the week was a bit busy during the evenings :-/

Reply:
After thinking a bit further, maybe it is an okay solution to use the crate's `Result` type after all.

Reply:
> Sorry about the delay, the week was a bit busy during the evenings :-/

That’s of course fine :)

On the the topic of errors, I'm not sure using the `rusqlite_migrations::Error` is the way to go here. There is a bunch of things in there that the hook  maybe we could simply have a message (in a `String`) in the `Result::Err()` (from the standard library), like this:
```diff
diff --git a/src/errors.rs b/src/errors.rs
index 1b6d450..0354db6 100644
--- a/src/errors.rs
+++ b/src/errors.rs
@@ -25,6 +25,8 @@ pub enum Error {
     MigrationDefinition(MigrationDefinitionError),
     /// The foreign key check failed
     ForeignKeyCheck(ForeignKeyCheckError),
+    /// Error returned by the hook
+    HookError(String)
 }
 
 impl Error {
@@ -51,6 +53,7 @@ fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
             Error::SpecifiedSchemaVersion(e) => Some(e),
             Error::MigrationDefinition(e) => Some(e),
             Error::ForeignKeyCheck(e) => Some(e),
+            Error::HookError(_) => None,
         }
     }
 }
diff --git a/src/lib.rs b/src/lib.rs
index c4a6722..ce41508 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -105,14 +105,14 @@
 };
 
 /// Helper trait to make hook functions clonable.
-pub trait MigrationHook: Fn(&Transaction) -> Result<()> + Sync {
+pub trait MigrationHook: Fn(&Transaction) -> std::result::Result<(), String> + Sync {
     /// Clone self.
     fn clone_box(&self) -> Box<dyn MigrationHook>;
 }
 
 impl<T> MigrationHook for T
 where
-    T: 'static + Clone + Sync + Fn(&Transaction) -> Result<()>,
+    T: 'static + Clone + Sync + Fn(&Transaction) -> std::result::Result<(), String>,
 {
     fn clone_box(&self) -> Box<dyn MigrationHook> {
         Box::new(self.clone())
@@ -427,7 +427,7 @@ fn goto_up(
             }
 
             if let Some(hook) = &m.up_hook {
-                hook(&tx)?;
+                hook(&tx).map_err(|e| Error::HookError(e))?;
             }
         }
```
 
And then it would be used like this:
```diff
diff --git a/src/tests.rs b/src/tests.rs
index 09c4207..cac29f4 100644
--- a/src/tests.rs
+++ b/src/tests.rs
@@ -235,7 +235,7 @@ fn hook_test() {
                     .unwrap();
                 }
 
-                Ok(())
+                Err("Failed hook".to_string())
             },
         ),
     ]);

```

I’ve also considered returning a fat pointer to an arbitrary error the hook would want to use, but it's hard to pass things around between threads then, because it adds other constraints to the error type.

What do you think?

Reply:
And I guess that return type could be named `HookResult` as you initially did, for convenience.

Reply:
> What do you think of passing an error message in a String?

Yes, sounds good. I wasn't sure if you were fine extending the error enum.

Reply:
I went slightly different and introduced a separate `HookError` which is eventually converted to `Error`. This avoids conflating concerns and because it implements `From<rusqlite::Error>` ergonomics are similar to other parts of the code. Hope this is okay.